spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/cassandra/CassandraMetricsAutoConfigurationTests.java

```diff
@@ -0,0 +1,168 @@
+			assertThat(registry.get(connectedNodes).gauge().value()).isEqualTo(7);
+			// Assert a typical Driver Timer metric
+			final String cqlRequests = getMetricName(context, "cql-requests");
+			assertThat(registry.find(cqlRequests).meter()).isNotNull();
+			assertThat(registry.get(cqlRequests).functionCounter().count()).isEqualTo(3);
+			// Assert timer snapshot metrics
+			final String cqlRequestsMean = getMetricName(context, "cql-requests.mean");
+			assertThat(registry.find(cqlRequestsMean).meter()).isNotNull();
+			assertThat(registry.get(cqlRequestsMean).functionCounter().count())
+					.isEqualTo(Duration.ofMillis(111).getNano());
+			final String cqlRequestsMin = getMetricName(context, "cql-requests.min");
+			assertThat(registry.find(cqlRequestsMin).meter()).isNotNull();
+			assertThat(registry.get(cqlRequestsMin).functionCounter().count())
+					.isEqualTo(Duration.ofMillis(3).getNano());
+			final String cqlRequestsMax = getMetricName(context, "cql-requests.max");
+			assertThat(registry.find(cqlRequestsMax).meter()).isNotNull();
+			assertThat(registry.get(cqlRequestsMax).functionCounter().count())
+					.isEqualTo(Duration.ofMillis(300).getNano());
+		});
+	}
+
+	/**
+	 * Cassandra driver metrics should be disabled if set to false via Spring metrics
+	 * management.
+	 */
+	@Test
+	void cassandraInstrumentationCanBeDisabledThroughSpringMetricsManagement() {
+		this.contextRunner.withBean(CqlSession.class, () -> mockSession(true))
+				.withPropertyValues("management.metrics.enable.spring.data.cassandra.driver.session=false")
+				.run((context) -> {
+					MeterRegistry registry = context.getBean(MeterRegistry.class);
+					assertThat(registry.find(getMetricName(context, "bytes-sent")).meter()).isNull();
+					assertThat(registry.find(getMetricName(context, "connected-nodes")).meter()).isNull();
+					assertThat(registry.find(getMetricName(context, "cql-requests")).meter()).isNull();
+					assertThat(registry.find(getMetricName(context, "cql-requests.mean")).meter()).isNull();
+				});
+	}
+
+	/**
+	 * Cassandra driver metrics should be disabled if no metrics are enabled in the
+	 * Driver's configuration.
+	 */
+	@Test
+	void cassandraInstrumentationIsDisabledIfNoMetricsEnabledInDriverConfig() {
+		this.contextRunner.withBean(CqlSession.class, () -> mockSession(false)).run((context) -> {
+			MeterRegistry registry = context.getBean(MeterRegistry.class);
+			assertThat(registry.find(getMetricName(context, "bytes-sent")).meter()).isNull();
+			assertThat(registry.find(getMetricName(context, "connected-nodes")).meter()).isNull();
+			assertThat(registry.find(getMetricName(context, "cql-requests")).meter()).isNull();
+			assertThat(registry.find(getMetricName(context, "cql-requests.mean")).meter()).isNull();
+		});
+	}
+
+	/**
+	 * Cassandra driver metrics should be disabled if {@link CqlSession} is not available
+	 * on the classpath.
+	 */
+	@Test
+	void cassandraInstrumentationIsDisabledIfDriverClassesNotOnClasspath() {
```

Comment:
I think that we should also test the case when `MetricsRegistry` AND `CqlSession` is not on the classpath. In that case the `MeterRegistry.getBean` should return null.

Reply:
It doesn't make sense to try to test with `CqlSession` filtered from the classpath. The ApplicationContextRunner tries to run with `CassandraMetricsAutoConfiguration`, which requires a `CqlSession` bean. If `CqlSession` isn't on the classpath, trying to spin up a context with `CassandraMetricsAutoConfiguration` will throw an exception indicating it can't find a required bean.

I can change the test so that it removes `MetricRegistry` from the classpath, but I'm mocking the `CqlSession` bean with a `MetricRegistry` since the tests don't have a fully initialized `CqlSession` bean with metrics configuration. I don't think removing `MetricRegistry` from the classpath at this point will prevent the metrics from being bound to the Micrometer `MeterRegistry`.

Or am I missing something?

Reply:
> It doesn't make sense to try to test with CqlSession filtered from the classpath. The ApplicationContextRunner tries to run with CassandraMetricsAutoConfiguration, which requires a CqlSession bean. If CqlSession isn't on the classpath, trying to spin up a context with CassandraMetricsAutoConfiguration will throw an exception indicating it can't find a required bean.

It shouldn't throw an exception, it should instead not be executed as part of the configuration process, since that class has a `@ConditionalOnClass({ MeterRegistry.class, CqlSession.class })` annotation.

This specific test does exactly that: it includes `CassandraMetricsAutoConfiguration` in the context, but excludes `CqlSession` from the classpath, to emulate the situation where the driver is not on the classpath.

I think what @tomekl007 wants is just another test similar to this one, but in which `MeterRegistry` would also be absent from the classpath, to emulate the situation where the user has excluded Micrometer _and_ the driver from the classpath.

I personally don't think that this is a common situation, but imo it doesn't hurt to add a test for that.

Reply:
I understand the intent. But I'm seeing the exception when I run the tests if I don't provide a CqlSession bean to the test's ApplicationContextRunner, i.e.

```java
this.contextRunner.withBean(CqlSession.class, () -> mockSession(true))
```
If I omit the bean, I get a NoSuchBeanDefinitionException with

```
No qualifying bean of type 'com.datastax.oss.driver.api.core.CqlSession' available.
```
Does it make sense to create the context with the mocked bean, but test without `CqlSession` on the classpath?

Reply:
I think I see the problem....

Reply:
Ok, with the latest changes, I can spin up the context with `CqlSession` and `MetricRegistry` missing from the classpath. The Micrometer metrics are indeed disabled, however the Micrometer `MeterRegisrty is not null, it simply has no meters in its map.